framework/calico_framework.py

```diff
@@ -0,0 +1,673 @@
+        """
+        Get all tasks across all tests.
+        """
+        for test in self.tests:
+            for task in test.tasks:
+                yield task
+
+    def all_tests_complete(self):
+        return len([test for test in self.tests if
+                    test.complete()]) == len(self.tests)
+
+    def unreserve_offers_by(self, test):
+        for slave_id, reserving_test in self.test_by_slave_id.iteritems():
+            if test == reserving_test:
+                del(self.test_by_slave_id[slave_id])
+
+
+    def kill_test(self, test, msg=None):
+        # Mark the test as done, and explain why
+        test.killed = True
+        test.additional_info = msg
+        if msg:
+            _log.error("Test Killed: %s. Reason: %s", test, msg)
+
+        # Unreserve any offers reserved by this test
+        self.unreserve_offers_by(test)
+
+        # If there's no other tests, we're done here
+        if self.all_tests_complete():
+            self.report_results_and_exit()
+
+    def registered(self, driver, frameworkId, masterInfo):
+        """
+        Callback used when the framework is succesfully registered.
+        """
+        _log.info("REGISTERED: with framework ID %s", frameworkId.value)
+
+    def resourceOffers(self, driver, offers):
+        """
+        Triggered when the framework is offered resources by mesos.
+        """
+        unreserved_offers_by_slave_id = {}
+
+        # Send new offers to tasks which reserved them, or add to unreserved
+        # offers pool
+        for offer in offers:
+            new_offer = Offer(offer)
+            if new_offer.slave_id in self.test_by_slave_id:
+                _log.info("New offer %s Sent to reserved slave", new_offer)
+                self.test_by_slave_id[new_offer.slave_id].launch({new_offer.slave_id: new_offer})
+            else:
+                _log.info("New offer %s Moved to unreserved pool", new_offer)
+                unreserved_offers_by_slave_id[new_offer.slave_id] = new_offer
+
+        # Offer unreserved offers to unstarted tests
+        if unreserved_offers_by_slave_id:
+            _log.info("Offering %d unreserved offers to Unstarted Tests", len(unreserved_offers_by_slave_id))
+            for test in self.tests:
+                if test.running() or test.complete():
+                    continue
+
+                _log.info("\t%s", test)
+                # Offer to the test
+                reserved_offers_by_slave_id = test.can_run_on(unreserved_offers_by_slave_id)
+                if reserved_offers_by_slave_id:
+                    # Save slave reservations for the future
+                    for slave_id in reserved_offers_by_slave_id:
+                        self.test_by_slave_id[slave_id] = test
+
+                    # Send reserved offers to slave
+                    offers_for_test = {}
+                    for slave_id, offer in reserved_offers_by_slave_id.iteritems():
+                        offers_for_test[slave_id] = offer
```

Comment:
isn't `offers_for_test` going to be identical to `reserved_offers_by_slave_id` at the conclusion of this loop?